unittests/test_meta.py

```diff
@@ -254,3 +256,91 @@ class AllowFinalOverride(Base):
 
         def foo(self):
             '''Overriding foo is now allowed.'''
+
+
+def test_callable_attributes(MyMeta):
+    '''Test issue #2113.
+
+    Setting a callable without the __name__ attribute would crash the
+    metaclass.
+    '''
+
+    class Callable:
+        def __call__(self):
+            pass
+
+    class Base(MyMeta):
+        f = Callable()
+
+
+def test_performance_function(MyMeta):
+    assert hasattr(MyMeta, '_rfm_perf_fns')
+
+    class Base(MyMeta):
+        @performance_function('units')
+        def perf_a(self):
+            pass
+
+        @performance_function('units')
+        def perf_b(self):
+            pass
+
+        def assert_perf_fn_return(self):
+            assert isinstance(
+                self.perf_a(), deferrable._DeferredPerformanceExpression
+            )
+
+    # Test the return type of the performance functions
+    Base().assert_perf_fn_return()
+
+    # Test the performance function set has been built correctly
+    assert len(Base._rfm_perf_fns) == 2
+    perf_fns = {'perf_a', 'perf_b'}
+    for fn in Base._rfm_perf_fns:
+        assert fn._rfm_perf_key in perf_fns
+        perf_fns.remove(fn._rfm_perf_key)
+
+    assert not perf_fns
+
+    class Derived(Base):
+        '''Test perf fn inheritance and perf_key argument.'''
+
+        def perf_a(self):
+            '''Override perf fn with a non perf fn.'''
+
+        @performance_function('units')
+        def perf_c(self):
+            '''Add a new perf fn.'''
+
+        @performance_function('units', perf_key='my_perf_fn')
+        def perf_d(self):
+            '''Perf function with custom key.'''
+
+    # Test the performance function set is correct with class inheritance
+    assert len(Derived._rfm_perf_fns) == 3
+    perf_fns = {'perf_b', 'perf_c', 'my_perf_fn'}
+    for fn in Derived._rfm_perf_fns:
+        assert fn._rfm_perf_key in perf_fns
+        perf_fns.remove(fn._rfm_perf_key)
+
+    assert not perf_fns
+
+
+def test_performance_function_errors(MyMeta):
+    with pytest.raises(ReframeSyntaxError):
```

Comment:
I have the feeling that here we can throw a `TypeError` and the error would still be traced back to the right user line. Similarly for anything in the `performance_function` decorator.

Reply:
> Changed to TypeError, but now I'm confused 😅. I thought the plan was to raise ReframeSyntaxError for everything that we impose and control.

If a programming error can be traced back to the user frame that triggered it, then ReFrame can avoid a stack trace automatically. In this case, it can be traced because it's a direct call chain from the decorator call. The problem has been with the variables and parameters, where we do several checks at a later point. In these cases, the user line information is lost and we can't trace back the error to the user code; the stack trace has no user frame. So to avoid printing a stack trace for such user errors, we invented the `ReframeSyntaxError`. That's the rule of thumb, I hope it makes sense...

Reply:
I don't know where would be the right place to be honest, because that's mostly a developer's documentation, isn't it?